smt.go

```diff
@@ -1,210 +1,392 @@
+	}
+
+	return currentHash, nil
+}
+
+// Get all the sibling nodes (sidenodes) for a given path from a given root.
+// Returns an array of sibling nodes, the leaf hash found at that path and the
+// leaf data. If the leaf is a placeholder, the leaf data is nil.
+func (smt *SparseMerkleTree) sideNodesForRoot(path []byte, root []byte) ([][]byte, []byte, []byte, error) {
+	sideNodes := make([][]byte, smt.depth())
+
+	if bytes.Equal(root, smt.th.placeholder()) {
+		// If the root is a placeholder, there are no sidenodes to return.
+		// Let the "actual path" be the input path.
+		return sideNodes, smt.th.placeholder(), nil, nil
+	}
+
+	currentData, err := smt.ms.Get(root)
+	if err != nil {
+		return nil, nil, nil, err
+	} else if smt.th.isLeaf(currentData) {
+		// If the root is a leaf, there are also no sidenodes to return.
+		return sideNodes, root, currentData, nil
+	}
+
+	var nodeHash []byte
+	for i := 0; i < smt.depth(); i++ {
+		leftNode, rightNode := smt.th.parseNode(currentData)
+
+		// Get sidenode depending on whether the path bit is on or off.
+		if hasBit(path, i) == right {
+			sideNodes[i] = leftNode
+			nodeHash = rightNode
+		} else {
+			sideNodes[i] = rightNode
+			nodeHash = leftNode
+		}
+
+		if bytes.Equal(nodeHash, smt.th.placeholder()) {
+			// If the node is a placeholder, we've reached the end.
+			return sideNodes, nodeHash, nil, nil
+		}
+
+		currentData, err = smt.ms.Get(nodeHash)
+		if err != nil {
+			return nil, nil, nil, err
+		} else if smt.th.isLeaf(currentData) {
+			// If the node is a leaf, we've reached the end.
+			break
+		}
+	}
+
+	return sideNodes, nodeHash, currentData, err
 }
 
 // Prove generates a Merkle proof for a key.
-func (smt *SparseMerkleTree) Prove(key []byte) ([][]byte, error) {
-    proof, err := smt.ProveForRoot(key, smt.Root())
-    return proof, err
+func (smt *SparseMerkleTree) Prove(key []byte) (SparseMerkleProof, error) {
+	proof, err := smt.ProveForRoot(key, smt.Root())
+	return proof, err
 }
 
 // ProveForRoot generates a Merkle proof for a key, at a specific root.
-func (smt *SparseMerkleTree) ProveForRoot(key []byte, root []byte) ([][]byte, error) {
-    sideNodes, err := smt.sideNodesForRoot(smt.digest(key), root)
-    return sideNodes, err
+func (smt *SparseMerkleTree) ProveForRoot(key []byte, root []byte) (SparseMerkleProof, error) {
+	path := smt.th.path(key)
+	sideNodes, leafHash, leafData, err := smt.sideNodesForRoot(path, root)
+
+	var nonEmptySideNodes [][]byte
+	for _, v := range sideNodes {
+		if v != nil {
+			nonEmptySideNodes = append(nonEmptySideNodes, v)
+		}
+	}
+
+	// Deal with non-membership proofs. If the leaf hash is the placeholder
+	// value, we do not need to add anything else to the proof.
+	var nonMembershipLeafData []byte
+	if !bytes.Equal(leafHash, smt.th.placeholder()) {
+		// This is a non-membership proof that involves showing a different leaf.
+		// Add the leaf data to the proof.
+		if err != nil {
+			return SparseMerkleProof{}, err
+		}
```

Comment:
Why is this err checked here? That is confusing, as it would occur several lines up in:
`sideNodes, leafHash, leafData, err := smt.sideNodesForRoot(path, root)`

Reply:
I'm not sure why I did that. Fixed.